File: fortran/src/mandelbrot.f90
module mandelbrot

    implicit none

    private
    public :: mandelbrot_set

contains

    function mandelbrot_set(grid_size, centre, range) result(set)

        implicit none

        integer,        intent(in) :: grid_size
        real, optional, intent(in) :: centre(2)
        real, optional, intent(in) :: range

        integer, parameter :: n_iters = 50
        integer            :: i, j, rank, ierr, comm_size
        real, allocatable  :: x_c(:), y_c_r(:), y_c(:), set(:,:)
        real               :: centre_l(2), range_l

        centre_l = [0.0, 0.0]
        if (present(centre)) centre_l = centre
        range_l = 4.0
        if (present(centre)) range_l = range

        ! Create grid arrays
        allocate(x_c(grid_size))
        allocate(y_c(grid_size))
        allocate(set(grid_size, grid_size))
        x_c = (/(i, i=1,grid_size, 1)/) * (2 * range_l / real(grid_size)) - range_l + centre_l(1)
        y_c = (/(i, i=1,grid_size, 1)/) * (2 * range_l / real(grid_size)) - range_l + centre_l(2)

        !$acc parallel loop
        do i = 1, size(x_c)
            do j = 1, size(y_c)
                if (converges(cmplx(x_c(i), y_c(j)), n_iters)) then
                    set(i, j) = 1.0
                else
                    set(i, j) = 0.0
                end if
            end do
        end do
        !$acc end parallel loop


    end function mandelbrot_set

    function converges(c, iters) result(conv_result)

        implicit none

        complex, intent(in) :: c
        integer, intent(in) :: iters
        logical             :: conv_result

        real, parameter :: threshold = 2.0
        integer         :: i
        complex         :: c_iter

        c_iter = (0,0)

        do i=0, iters
            c_iter = c_iter**2 + c
            if (abs(c_iter) > threshold) then
                conv_result = .false.
                return
            end if
        end do

        conv_result = .true.

        return

    end function converges

end module mandelbrot

File: rust/src/mandelbrot.rs
use std::iter::FromIterator;
use num::complex::{Complex, ComplexFloat};
use ndarray::Array2;

pub fn mandelbrot_set(res:usize, centre:[f32; 2], range:f32) -> Array2::<f32> {

    let n_iters = 50;
    let res_factor = range / (res as f32);

    let x_c = Vec::from_iter((0..res).map(|i| ((i as f32) * res_factor) -
                                                centre[0] - (range / 2.0)));
    let y_c = Vec::from_iter((0..res).map(|i| ((i as f32) * res_factor) -
                                                centre[1] - (range / 2.0)));
    let mut set = Array2::<f32>::zeros((res, res));

    for i in 0..res {
        for j in 0..res {
            if converges(Complex::new(x_c[i], y_c[j]), n_iters) {
                set[[i, j]] = 1.0;
            } else {
                set[[i, j]] = 0.0;
            }
        }
    }

    return set;
}

fn converges(c:Complex<f32>, iters:usize) -> bool {

    let threshold:f32 = 2.0;

    let mut c_iter = Complex::new(0f32, 0f32);
    for _ in 0..iters {
        c_iter = c_iter.powf(2.0) + c;
        if c_iter.abs() > threshold {
            let conv_result = false;
            return conv_result
        }
    }

    let conv_result = true;

    return conv_result;
}